### ISSUE TITLE
RFC: anonymous usage statistics

### ISSUE THREAD

**Issue by ktsaou** [netdata/netdata#4735]
## Motivation

We need anonymous statistics about netdata. Such statistics will be used for:

1. **Quality assurance**, to help us understand if netdata behaves as expected and help us identify repeating issues for certain distributions or environment.

2. **Usage statistics**, to help us focus on the parts of netdata that are used the most, or help us identify the extend our development decisions influence the community (i.e. can we drop a certain collector? which is the most used collector to optimize? can we drop node.d.plugin? is anyone using netdata statsd? which backends are used by how many? etc).

## Collection points

These are the points we can collect these statistics:

1. The netdata installer, updater, uninstaller. These reports will not exist when we will distribute binary packages.
2. In the unlikely event netdata crashes.
3. When netdata starts (1 minute after it starts, to report also plugins and modules collecting data).
4. When a user accesses a netdata dashboard.

## Opt out

Users should be able to opt-out from this data collection.

To make it easier for users to opt-out globally, they will need to create the file `/etc/netdata/.opt-out-from-anonymous-statistics`, with one of following values:

- `all`, to opt-out from all kinds of statistics, including crashes
- `usage`, to opt-out from usage statistics only (points 3 and 4 above)
- `user`, to opt-out from user usage statistics only (point 4 above)

This file will be generated by the custom installer scripts, with the command line option `--out-out-from-anonymous-statistics VALUE`, so that a user that is not willing to share anything, will give this option to the installer and no data will be shared at all.

## Public reports

We will publish reports with these data. All data will be public, for everyone to see and examine, with the exception of 2 values:

- The `IPs` that sent the usage statistics
- The unique netdata machine IDs that sent the usage statistics

The above values will be used in calculating statistics, but will be deleted from the database as soon as the calculations of the statistics allow.

## Information to be collected

We should maintain a public wiki, explaining the usage statistics we maintain, the purpose we collect it for and the actual reports we generated using these statistics.

All this information will be public.

### netdata info

The following information about netdata itself:

- version as given by `netdata -V`

- unique netdata machine ID (random, generated when netdata starts) as stored in `/var/lib/netdata/registry/netdata.public.unique.id`

### operating system info

- The following information from `/etc/os-release`:
  - `NAME`, like `Debian GNU/Linux`, `Gentoo`, `Manjaro Linux`
  - `VERSION_ID`, like `9`
  - `ID`, like `debian`, `gentoo`, `manjaro`
  - `ID_LIKE`, like `arch`

- The following information from `uname`
  - Kernel name, as given by `uname -s`, example `Linux`
  - Kernel version, as given by `uname -r`, example `4.19.2-1-MANJARO`
  - Architecture, as given by `uname -m`, example `x86_64`

- Information about the virtualization technology
  - The output of the command `systemd-detect-virt`, example `none`, or `kvm`

### Installation reports

Installation reports (install, update, uninstall) should also report the final status of the operation (`OK`, `FAILED`, `CANCELLED`).

### Crash reports

Netdata daemon crash reports should provide a stack trace when possible.

## Implementation

We could provide a script that will be installed as `/usr/libexec/netdata/plugins.d/send-anonymous-usage-info.sh` to take care of all reports, except the user one (point 4).

This script will collect and send all reports, respecting the contents of `/etc/netdata/.opt-out-from-anonymous-statistics`.

This script will be called by all installation scripts and netdata itself to send the information required by points 1, 2 and 3.

Netdata itself should read the contents of `/etc/netdata/.opt-out-from-anonymous-statistics` and expose this information to the dashboard, so that the dashboard will not send user statistics (point 4).

**Comment by paulfantom** [comment]
This is largely a duplicate of #4041

**Comment by paulfantom** [comment]
Generally yes, with some minor caveats:
- IP and shouldn't be collected at all (it shouldn't even go over the wire) to be compliant with GDPR. If this is necessary, then this system cannot be opt-out and needs to be opt-in.
- `systemd-detect-virt` isn't very portable or common. We cannot use it inside docker container or on non-systemd hosts.
- I don't see any value in installation/deinstallation reports. If everything went ok, then we don;t need such thing and if sth went wrong we usually get an issue opened.


Also this is still a duplicate so for the sake of clarity discussion should be happening in original thread.

**Comment by ktsaou** [comment]
> * IP and shouldn't be collected at all (it shouldn't even go over the wire) to be compliant with GDPR. If this is necessary, then this system cannot be opt-out and needs to be opt-in.

IPs cannot be avoided. They will be known to our server, not matter what we do. We will not use them, apart for security (spam protection, etc). They could also be a strong indication that we really have that many servers installed (i.e. if 50% of our installations are reported by single IP, something is very wrong).

GDPR is about personal data. Any data (including IPs) that are not personalized are outside the scope of GDPR (yes, I am sure).

> * `systemd-detect-virt` isn't very portable or common. We cannot use it inside docker container or on non-systemd hosts.

hm... There is a function called `iscontainer()` in functions.sh. We could use that with some adaptions.

> * I don't see any value in installation/deinstallation reports. If everything went ok, then we don;t need such thing and if sth went wrong we usually get an issue opened.

We need these to know failures to install and their ratio. So, if we have 100 failures to install on distro X, is that bad? It may be if we have none successful - it may not be if we have 10000 successful.

They will also be useful for monitoring auto-updates. Like: suddenly all debian machines failed to auto-update netdata...

So, these are QA of our installation scripts. They will be obsolete when we will have our own packaging.

**Comment by cakrit** [comment]
I agree of course that we need to be able to collect data, but data collection (even anonymous) must be opt-in, not opt-out. We've had some strong, understandable complaints in #2760. Trust is paramount here, all data collection should be disabled by default, even if it hurts. 

What I'm not sure about is when/where the question to opt-in should appear. Upon first use of the web GUI perhaps.

**Comment by paulfantom** [comment]
I am against of telemetry as an opt-in feature for following reasons:
1. We don't have any way to modify netdata configuration from WebGUI and that is good from security perspective. This means that users need to modify config files, but on the other hand we advertise that netdata uses autoconfiguration and there is not much need to touch config files.
2. Installation from binary packages doesn't allow you to ask any questions during installation process, so we cannot ask if user want to participate in telemetry.
3. Majority of users won't care about sending data to us as this won't have any pros for them

Basically if we apply telemetry as an opt-in then we will have only a small fraction of data which won't be representative. This way we will implement a feature which will tell us which parts of netdata are used, but this won't be statistically meaningful.

**Comment by ktsaou** [comment]
I agree with both of you, but I would prefer the @paulfantom 	way (enabled by default, action required additional action to opt-out), since this is the only way to have practical statistics.

Of course, this would require clear documentation about it (all over the place and centralized at a wiki) and simple actions to disable it.

Since netdata is free and open-source software, having such minimal anonymous usage statistics is a reasonable trade off.

The term `anonymous` is key here. We should not retrieve and store anything that could be considered sensitive by anyone (i.e. no URLs, no hostnames - the IP will be known because there is no way to hide it, but it will not be stored or processed, apart for security reasons - btw according to GDPR these are not sensitive data if we don't know to whom they belong - but still people feel uncomfortable - totally understandable).

In #2760 	folks considered the hostname and URL as sensitive info. If these were not collected, I think we wouldn't have any issues.

**Comment by ktsaou** [comment]
Example script to do the job:

```sh
#!/usr/bin/env sh

# Valid actions:
# - KICKSTART   - installed using the kickstart.sh script
# - KICKSTART64 - installed using the kickstart-static64.sh script
# - MAKESELF    - installed by running a makeself package
# - INSTALL     - installed using netdata-installer.sh
# - UPDATE      - updated using netdata-updater.sh
# - FATAL       - netdata exited due to a fatal condition
# - START       - netdata started

ACTION="${1}"
ACTION_RESULT="${2}"
ACTION_DATA="${3}"

# -------------------------------------------------------------------------------------------------
# detect the operating system

OS_DETECTION=
NAME=
VERSION=
VERSION_ID=
ID=
ID_LIKE=

if [ -f "/etc/os-release" ]; then
    OS_DETECTION="/etc/os-release"
    eval "$(cat /etc/os-release | grep -E "^(NAME|ID|ID_LIKE|VERSION|VERSION_ID)=")"
elif [ -f "/etc/lsb-release" ]; then
    OS_DETECTION="/etc/lsb-release"
    DISTRIB_ID= DISTRIB_RELEASE= DISTRIB_CODENAME=
    eval "$(cat /etc/lsb-release | grep -E "^(DISTRIB_ID|DISTRIB_RELEASE|DISTRIB_CODENAME)=")"
    NAME="${DISTRIB_ID}"
    VERSION="${DISTRIB_RELEASE}"
    ID="${DISTRIB_CODENAME}"
elif [ ! -z "$(which lsb_release 2>/dev/null || command -v lsb_release 2>/dev/null)" ]; then
    OS_DETECTION="lsb_release"
    NAME="$(lsb_release -i 2>/dev/null | sed "s/[[:space:]]*:[[:space:]]*/:/g" | cut -d ':' -f 2)"
    VERSION_ID="$(lsb_release -f 2>/dev/null | sed "s/[[:space:]]*:[[:space:]]*/:/g" | cut -d ':' -f 2)"
    VERSION="$(lsb_release -c 2>/dev/null | sed "s/[[:space:]]*:[[:space:]]*/:/g" | cut -d ':' -f 2)"
fi

# -------------------------------------------------------------------------------------------------
# detect the kernel

KERNEL_NAME="$(uname -s)"
KERNEL_VERSION="$(uname -r)"
ARCHITECTURE="$(uname -m)"

# -------------------------------------------------------------------------------------------------
# detect the virtualization

VIRTUALIZATION="unknown"
if [ ! -z "$(which systemd-detect-virt 2>/dev/null || command -v systemd-detect-virt 2>/dev/null)" ]
then
    VIRTUALIZATION="$(systemd-detect-virt)"
else
    # /proc/1/sched exposes the host's pid of our init !
    # http://stackoverflow.com/a/37016302
    pid=$( cat /proc/1/sched 2>/dev/null | head -n 1 | { IFS='(),#:' read name pid th threads; echo $pid; } )
    if [ ! -z "${pid}" ]; then
        pid=$(( pid + 0 ))
        [ ${pid} -ne 1 ] && VIRTUALIZATION="container"
    fi

    # ubuntu and debian supply /bin/running-in-container
    # https://www.apt-browse.org/browse/ubuntu/trusty/main/i386/upstart/1.12.1-0ubuntu4/file/bin/running-in-container
    [ -x "/bin/running-in-container" ] && "/bin/running-in-container" >/dev/null 2>&1 && VIRTUALIZATION="container"

    # lxc sets environment variable 'container'
    [ ! -z "${container}" ] && VIRTUALIZATION="lxc"

    # docker creates /.dockerenv
    # http://stackoverflow.com/a/25518345
    [ -f "/.dockerenv" ] && VIRTUALIZATION="docker"
fi

# -------------------------------------------------------------------------------------------------
# check netdata version

if [ -z "${NETDATA_VERSION}" ]; then
    NETDATA_VERSION="$(netdata -V 2>&1 | cut -d ' ' -f 2)"
fi

if [ -z "${NETDATA_REGISTRY_UNIQUE_ID}" ]; then
    for dir in "$(netdata -W get global 'lib directory' '/var/lib/netdata' 2>/dev/null)" "/var/lib/netdata" "/opt/netdata/var/lib/netdata"; do
        if [ -f "${dir}/registry/netdata.public.unique.id" ]; then
            NETDATA_REGISTRY_UNIQUE_ID="$(cat "${dir}/registry/netdata.public.unique.id")"
            break
        fi
    done
fi

OPTOUT=
if [ -z "${NETDATA_CONFIG_DIR}" ]; then
    for dir in "$(netdata -W get global 'config directory' '/etc/netdata' 2>/dev/null)" "/etc/netdata" "/opt/netdata/etc/netdata"; do
        if [ -f "${dir}/.opt-out-from-anonymous-statistics" ]; then
            OPTOUT="$(cat "${dir}/.opt-out-from-anonymous-statistics")"
            break
        fi
    done
else
    OPTOUT="$(cat "${NETDATA_CONFIG_DIR}/.opt-out-from-anonymous-statistics")"
fi

if [ "${OPTOUT}" = "all" ] || [ "${OPTOUT}" = "usage" -a "${ACTION}" = "START" ]; then
    # the user does not want us to send statistics
    exit 0
fi

# -------------------------------------------------------------------------------------------------
# send the anonymous statistics to netdata

# curl -Ss >/dev/null 2>&1 --max-time 5
echo "https://registry.my-netdata.io/log/anonymous-statistics?v=1&version=${NETDATA_VERSION}&machine_guid=${NETDATA_REGISTRY_UNIQUE_ID}&os_detection=${OS_DETECTION}&distro_name=${NAME}&distro_id=${ID}&distro_id_like=${ID_LIKE}&distro_version=${VERSION}&distro_version_id=${VERSION_ID}&kernel_name=${KERNEL_NAME}&kernel_version=${KERNEL_VERSION}&architecture=${ARCHITECTURE}&virtualization=${VIRTUALIZATION}"

```

**Comment by ktsaou** [comment]
Created WIP PR #4743

**Comment by Ferroin** [comment]
> IPs cannot be avoided. They will be known to our server, not matter what we do. We will not use them, apart for security (spam protection, etc).

Disclosure can't be avoided, but there's no reason we need to log them (or if we do log them, that we have to retain them long term).

> They could also be a strong indication that we really have that many servers installed (i.e. if 50% of our installations are reported by a single IP, something is very wrong).

Except this isn't reliable for the same reasons that IP's aren't PII.  There are actual, currently operating, ISP's who have thousands of customers in a NAT routing configuration behind a _single_ IP address.

**Comment by paulfantom** [comment]
> ISP's who have thousands of customers in a NAT routing configuration behind a single IP address

Basically half of Great Britain :smile:

**Comment by cakrit** [comment]
Since we agreed we'll be using GA for this:
- The script will be replaced by one that uses the [GA measurement protocol](https://developers.google.com/analytics/devguides/collection/protocol/v1/)
- anonymizeip will be set to true, so that GA doesn't get the full IP
- we will be sending a fake hostname
- we will need to define a few additional meaningful events that GA will have a hard time extrapolating from the data. e.g. "Frequent usage" or "Auto Update Enabled"

**Comment by cakrit** [comment]
PR #4743 replaced by #5113 and is in progress

**Comment by cakrit** [comment]
We discussed a lot with Pawel on how this will work for the various installation methods, most of which actually call the `netdata-installer.sh`. We agreed that he'll make a change to add a file `/var/lib/netdata/.installation` that will contain the installation method used regardless of the case we fall into. The anonymous statistics script will be reading that file whenever it is executed with action START, to check if it exists. if it does, it will trigger the install action with the proper installation method in addition to the START action and then remove that file.

So the various installers won't be calling the statistics script themselves. 

I'll just need to see how makeself will do something similar.

**Comment by cakrit** [comment]
We also agreed with @paulfantom that it will be a bit confusing to the users and not that useful to have two types of opt-out (all vs usage), plus it will make the installation options slightly more tricky. So we'll just keep it simple and only require the existence of /etc/netdata/.opt-out-from-anonymous-statistics to prevent sending info to GA.

**Comment by cakrit** [comment]
After a call with @ktsaou, we decided to only implement the netdata start, stop, fatal and the UI at this point.

**Comment by cakrit** [comment]
To make the information public, we can use the [GA Embed API](https://ga-dev-tools.appspot.com/embed-api/basic-dashboard/).
